File: jets/userflow/validate.go
// Reference checks over a user flow document, and the deployment-time switch
// that decides how severely one of them is taken.
//
// This is the port of jetsclient_ide/src/userflow/validate.ts, kept literally
// parallel to it: same finding codes, same order, same messages. Two
// implementations of one rule stay in agreement only if a reader can put them
// side by side, so structural improvements to one belong in both or in neither.
//
// **What is here and what is still S.4's.** These are the checks a JSON Schema
// cannot express — schema.ts absorbs everything expressible in the document's
// shape. Wiring them into SaveWorkspaceFileContent, beside the existing
// well-formed-JSON check, is S.4's task and is deliberately not done here.
package userflow

import (
	"encoding/json"
	"fmt"
	"os"
	"slices"
	"strings"
)

// Severity of a finding. A document with any error must not be saved; warnings
// are reported and do not block.
type Severity string

const (
	Error   Severity = "error"
	Warning Severity = "warning"
)

// Finding codes, matching validate.ts.
const (
	CodeUnknownStartState = "unknownStartState"
	CodeUnknownTarget     = "unknownTarget"
	CodeUnreachableState  = "unreachableState"
)

type Finding struct {
	Severity Severity `json:"severity"`
	Code     string   `json:"code"`
	Message  string   `json:"message"`
	// Path locates the offending place as a JSON Pointer (RFC 6901) into the
	// document — "/states/select_client/choices/0/nextState". Empty means the
	// finding is about the document as a whole.
	//
	// **Added at the agentic_ai stream's request, and it was right that it helps
	// this side sooner.** The structure was already here and was being
	// stringified into Message: an unknownTarget knows which state and which
	// transition offended, and putting that only in prose means a UI cannot jump
	// to it and a repair prompt cannot say where. One string rather than a
	// typed location keeps Finding comparable and JSON-friendly, and keeps the
	// two file types from having to agree on a schema for locations.
	//
	// No escaping is applied and none is needed: every segment that varies is an
	// Identifier, whose pattern excludes both "/" and "~" (schema.ts). If a
	// future segment can contain either, it must be escaped per RFC 6901 §3
	// rather than interpolated.
	Path string `json:"path,omitempty"`
}

// Policy says how severely each configurable check is taken. A struct rather
// than a bool so that a second configurable check does not change every
// signature.
type Policy struct {
	UnreachableState Severity
}

// StrictReachabilityEnvVar is the deployment-time switch.
//
// **The trade this was introduced with turned out not to exist, and the entry
// that describes it is I-18.** It was added believing that pipelineConfigUF
// shipped two dead states, so that a strict deployment could not save a flow it
// already runs. Those two states are reached by a button — an action that jumps
// the flow — and the walk could not see the edge because the document did not
// declare it. It does now, via State.GoToStates, and **no flow in the corpus is
// refused under the strict policy**.
//
// The default stays a warning regardless, because the two questions are
// different: whether the shipping flows pass is about today, and whether a
// workspace wants an unreached state to block a save is about that workspace.
// A deployment that wants the stricter reading can now take it without also
// taking a false positive.
const StrictReachabilityEnvVar = "JETS_USERFLOW_STRICT_REACHABILITY"

// DefaultPolicy reports unreachable states without blocking the save.
func DefaultPolicy() Policy { return Policy{UnreachableState: Warning} }

// StrictPolicy refuses a document with a state nothing transitions to.
func StrictPolicy() Policy { return Policy{UnreachableState: Error} }

// PolicyFromEnv resolves the policy from the process environment.
//
// Truthy is "1", "true", "yes" or "on", case-insensitive and trimmed. Anything
// else, including an unset variable, leaves the warning a warning — so a typo
// in the value fails safe rather than silently turning enforcement on.
func PolicyFromEnv() Policy {
	if IsTruthy(os.Getenv(StrictReachabilityEnvVar)) {
		return StrictPolicy()
	}
	return DefaultPolicy()
}

// IsTruthy matches isTruthy in validate.ts. Kept exported and separate for that
// reason: it is the one place the two implementations must agree character for
// character, and a shared test table is cheaper than a shared reading.
func IsTruthy(value string) bool {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

// Choice is a guarded transition. Only the target is read here; the condition is
// the schema's business and the interpreter's.
type Choice struct {
	NextState string `json:"nextState"`
}

// State carries only the fields the reference checks need. Unknown fields are
// ignored on purpose — the schema check runs first and is what rejects them, so
// duplicating that here would give one document two different complaints.
//
// GoToStates are transitions an action of this state makes rather than ones the
// Next button makes. They are edges of the state machine all the same, and
// omitting them is what made S.1 report two live states as dead — see I-18 and
// the comment on targetsOf.
type State struct {
	IsEnd            bool     `json:"isEnd"`
	Choices          []Choice `json:"choices"`
	DefaultNextState string   `json:"defaultNextState"`
	GoToStates       []string `json:"goToStates"`
}

type Flow struct {
	StartAtKey string           `json:"startAtKey"`
	States     map[string]State `json:"states"`
}

// ParseFlow reads a .uf.json document. It does not validate against the schema;
// that is the caller's first step and a separate library's job.
func ParseFlow(data []byte) (*Flow, error) {
	var flow Flow
	if err := json.Unmarshal(data, &flow); err != nil {
		return nil, fmt.Errorf("user flow is not valid JSON: %w", err)
	}
	return &flow, nil
}

// Transition is a target and where in the document it was declared.
type Transition struct {
	Target string
	Path   string
}

// targetsOf returns every state a state transitions to: choices, then the
// default, then the ones an action takes — the same order as validate.ts, so
// findings come out in the same order from both.
func targetsOf(flow *Flow, key string) []Transition {
	state, ok := flow.States[key]
	if !ok {
		return nil
	}
	base := fmt.Sprintf("/states/%s", key)
	targets := make([]Transition, 0, len(state.Choices)+len(state.GoToStates)+1)
	for i, choice := range state.Choices {
		targets = append(targets, Transition{
			Target: choice.NextState,
			Path:   fmt.Sprintf("%s/choices/%d/nextState", base, i),
		})
	}
	if state.DefaultNextState != "" {
		targets = append(targets, Transition{
			Target: state.DefaultNextState,
			Path:   base + "/defaultNextState",
		})
	}
	for i, target := range state.GoToStates {
		targets = append(targets, Transition{
			Target: target,
			Path:   fmt.Sprintf("%s/goToStates/%d", base, i),
		})
	}
	return targets
}

// ValidateFlow runs the reference checks over a document that has already passed
// the schema.
//
// Order is deliberate: an unknown start state suppresses the reachability walk,
// because a walk from nowhere reports every state as unreachable and buries the
// one finding that matters.
func ValidateFlow(flow *Flow, policy Policy) []Finding {
	findings := make([]Finding, 0)

	_, startExists := flow.States[flow.StartAtKey]
	if !startExists {
		findings = append(findings, Finding{
			Severity: Error,
			Code:     CodeUnknownStartState,
			Message:  fmt.Sprintf("startAtKey %q is not a state", flow.StartAtKey),
			Path:     "/startAtKey",
		})
	}

	// Sorted so that two runs over one document report in one order. Go's map
	// iteration is deliberately randomised, and a validator whose output moves
	// between runs is one nobody can diff.
	keys := sortedKeys(flow.States)
	for _, key := range keys {
		for _, transition := range targetsOf(flow, key) {
			if _, ok := flow.States[transition.Target]; !ok {
				findings = append(findings, Finding{
					Severity: Error,
					Code:     CodeUnknownTarget,
					Message: fmt.Sprintf(
						"state %q transitions to %q, which is not a state", key, transition.Target),
					Path: transition.Path,
				})
			}
		}
	}

	if startExists {
		reached := map[string]bool{flow.StartAtKey: true}
		frontier := []string{flow.StartAtKey}
		for len(frontier) > 0 {
			key := frontier[len(frontier)-1]
			frontier = frontier[:len(frontier)-1]
			for _, transition := range targetsOf(flow, key) {
				if _, ok := flow.States[transition.Target]; ok && !reached[transition.Target] {
					reached[transition.Target] = true
					frontier = append(frontier, transition.Target)
				}
			}
		}
		for _, key := range keys {
			if !reached[key] {
				findings = append(findings, Finding{
					Severity: policy.UnreachableState,
					Code:     CodeUnreachableState,
					Message: fmt.Sprintf(
						"state %q is not reachable from %q", key, flow.StartAtKey),
					Path: fmt.Sprintf("/states/%s", key),
				})
			}
		}
	}

	return findings
}

// ErrorsOnly is what a save-time check acts on; warnings are reported alongside.
func ErrorsOnly(findings []Finding) []Finding {
	errs := make([]Finding, 0, len(findings))
	for _, f := range findings {
		if f.Severity == Error {
			errs = append(errs, f)
		}
	}
	return errs
}

func sortedKeys(states map[string]State) []string {
	keys := make([]string, 0, len(states))
	for key := range states {
		keys = append(keys, key)
	}
	slices.Sort(keys)
	return keys
}

File: jetsclient_ide/src/userflow/validate.ts
/**
 * The rules a JSON Schema cannot state. Task S.1.
 *
 * `schema.ts` absorbs everything expressible in the document's *shape*,
 * including two rules the Dart validator either checks or only documents. What
 * is left here is reference integrity, which JSON Schema has no vocabulary for:
 * a `$ref` points at a schema, never at a sibling key's value.
 *
 * **S.4 ports this file to Go**, beside the schema check in
 * `SaveWorkspaceFileContent`. It is written to make that port mechanical — no
 * dependency on zod, plain data in and a list of findings out, and the same
 * shape of return the Dart's `validateConfiguration()` has
 * (`user_flow_config.dart:61`).
 *
 * ## What the Dart checks, and the one it misses
 *
 * `validateConfiguration()` checks three things: the start state exists, each
 * map key matches its state's `key` field, and a non-end state has somewhere to
 * go. The second disappears in this representation — a state has no `key` field
 * to disagree with, because `states` is keyed by it. The third is structural in
 * `schema.ts`.
 *
 * **It never checks that a transition names a state that exists**, and the
 * failure it lets through is not a diagnostic: `user_flow_actions.dart:82` does
 * `states[nextStateKey]!.formConfig` and the `!` turns a typo into a crash on
 * the button press. That check is `unknownTarget` below, and it is the one rule
 * here the Flutter app does not have.
 */

import type { UserFlow } from "./schema";

export type Severity = "error" | "warning";

/**
 * How severely each finding is taken. One entry today; a struct rather than a
 * boolean so a second configurable check does not change every call site.
 *
 * **The default is the shipping corpus's answer, not a preference.**
 * `pipelineConfigUF` ships two unreachable states, so a deployment that took
 * them as errors could not save the flows it already runs.
 */
export interface Policy {
  unreachableState: Severity;
}

export const defaultPolicy: Policy = { unreachableState: "warning" };
export const strictPolicy: Policy = { unreachableState: "error" };

/**
 * The deployment-time switch, resolved from an environment.
 *
 * **`JETS_USERFLOW_STRICT_REACHABILITY` is read by the server, not by the
 * browser**, and that is the whole reason this function takes its environment as
 * an argument rather than reaching for `process.env` or `import.meta.env`. A
 * bundle cannot carry a deployment-time value: vite bakes `import.meta.env` in
 * at build time — the same property that made `/ide/` non-relocatable in A.1 —
 * and a browser has no environment to read at run time.
 *
 * So the authority is Go, at the save-time check the gate settled on (G.3):
 * `jets/userflow/policy.go` reads the variable and its `ValidateFlow` obeys it.
 * This side exists for the two consumers that *do* have an environment — the
 * tests, and any build-time or CLI validation — and to keep the two
 * implementations literally parallel, which is what makes them stay in
 * agreement.
 *
 * **The IDE's own copy is advisory until a channel exists to tell it.** Nothing
 * ships the policy to the browser today, so under a strict deployment an author
 * would see a warning and then have the save refused. That is the right way
 * round — the server is the enforcement point and a client check is presentation
 * (assessment §3.5) — but it is a worse message than it needs to be, and S.4
 * should send the policy along with the flow.
 *
 * Truthy is `1`, `true`, `yes` or `on`, case-insensitive and trimmed. Anything
 * else, including an unset variable, leaves the warning a warning.
 */
export function policyFromEnv(env: Record<string, string | undefined>): Policy {
  return isTruthy(env["JETS_USERFLOW_STRICT_REACHABILITY"]) ? strictPolicy : defaultPolicy;
}

/** Kept separate so the Go port has one obvious thing to match. */
export function isTruthy(value: string | undefined): boolean {
  if (value === undefined) return false;
  switch (value.trim().toLowerCase()) {
    case "1":
    case "true":
    case "yes":
    case "on":
      return true;
    default:
      return false;
  }
}

export interface Finding {
  severity: Severity;
  /** Machine-readable so the Go port and the IDE can agree on a message. */
  code: "unknownStartState" | "unknownTarget" | "unreachableState";
  message: string;
}

/** Every state a state can transition to, choices first then the default. */
export function targetsOf(flow: UserFlow, stateKey: string): string[] {
  const state = flow.states[stateKey];
  if (state === undefined) return [];
  const targets = "choices" in state && state.choices ? state.choices.map((c) => c.nextState) : [];
  if ("defaultNextState" in state && state.defaultNextState !== undefined) {
    targets.push(state.defaultNextState);
  }
  return targets;
}

/**
 * Reference checks over a document that has already passed the schema.
 *
 * Order is deliberate: an unknown start state suppresses the reachability walk,
 * because a walk from nowhere reports every state as unreachable and buries the
 * one finding that matters.
 */
export function validateFlow(flow: UserFlow, policy: Policy = defaultPolicy): Finding[] {
  const findings: Finding[] = [];
  const keys = Object.keys(flow.states);

  if (!(flow.startAtKey in flow.states)) {
    findings.push({
      severity: "error",
      code: "unknownStartState",
      message: `startAtKey "${flow.startAtKey}" is not a state`,
    });
  }

  for (const key of keys) {
    for (const target of targetsOf(flow, key)) {
      if (!(target in flow.states)) {
        findings.push({
          severity: "error",
          code: "unknownTarget",
          message: `state "${key}" transitions to "${target}", which is not a state`,
        });
      }
    }
  }

  if (flow.startAtKey in flow.states) {
    const reached = new Set<string>([flow.startAtKey]);
    const frontier = [flow.startAtKey];
    while (frontier.length > 0) {
      for (const target of targetsOf(flow, frontier.pop()!)) {
        if (!reached.has(target) && target in flow.states) {
          reached.add(target);
          frontier.push(target);
        }
      }
    }
    for (const key of keys) {
      if (!reached.has(key)) {
        findings.push({
          severity: policy.unreachableState,
          code: "unreachableState",
          message: `state "${key}" is not reachable from "${flow.startAtKey}"`,
        });
      }
    }
  }

  return findings;
}

/**
 * **Unreachability is a warning by default, and the corpus is the reason.**
 *
 * `pipelineConfigUF` ships with two states nothing transitions to —
 * `add_merge_process_inputs` and `add_injected_process_inputs`, sitting under a
 * comment reading "new_process_input is implemented as a dialog". They are the
 * remains of an approach that was replaced, and the two dialog forms that
 * replaced them (`pcNewProcessInputDialog`, `pcNewProcessInputDialog4MI`) are
 * among the four registered forms no flow state references.
 *
 * So a reachability *error* would reject a configuration that has shipped for
 * years and works — which is precisely the case the plan's rule covers: a real
 * config that fails means the check is wrong, not the config. It stays as a
 * warning, where it is still worth having: a state nothing reaches is either
 * dead or a typo in the transition that should have reached it, and both are
 * worth a line in the IDE.
 *
 * A second entry point would change this analysis, and one exists — the
 * `startAtKey` navigation parameter (`screens/user_flow_screen.dart:105`) lets a
 * screen enter a flow anywhere. It does not rescue these two: the three screens
 * that use it name `select_client`, `select_source_config` and
 * `select_pipeline_config` (`modules/screen_config_impl.dart`). If entry points
 * ever become part of the document, this walk should start from all of them.
 *
 * **A deployment that wants it as an error can have one**, through
 * `JETS_USERFLOW_STRICT_REACHABILITY` — see `policyFromEnv` above. The default
 * is what it is because of two states in one flow, not because the check is
 * weak: a workspace with no such history has no reason to tolerate a state
 * nothing reaches, and turning it up costs a variable. **A deployment that sets
 * it cannot save `pipelineConfigUF` unmodified**, which is the trade being
 * bought, and which should be found out at deploy time rather than at save time.
 */
export const errorsOnly = (findings: Finding[]): Finding[] =>
  findings.filter((f) => f.severity === "error");
